clang-tools-extra/clang-tidy/bsl/StructDefCheck.cpp

```diff
@@ -0,0 +1,128 @@
+//===--- StructDefCheck.cpp - clang-tidy ----------------------------------===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#include "StructDefCheck.h"
+#include "clang/AST/ASTContext.h"
+#include "clang/ASTMatchers/ASTMatchFinder.h"
+
+#include <iostream>
+
+using namespace clang::ast_matchers;
+
+namespace clang {
+namespace tidy {
+namespace bsl {
+
+AST_MATCHER(CXXRecordDecl, hasBases) {
+  return Node.getNumBases() > 0 || Node.getNumVBases() > 0;
+}
+
+void StructDefCheck::registerMatchers(MatchFinder *Finder) {
+  Finder->addMatcher(
+      cxxMethodDecl(
+          isUserProvided(),
+          anyOf(cxxConstructorDecl(unless(anyOf(isDefaulted(), isDeleted()))),
+                cxxDestructorDecl(), isCopyAssignmentOperator(),
+                isMoveAssignmentOperator()))
+          .bind("special"),
+      this);
+  Finder->addMatcher(
+      cxxMethodDecl(
+          isUserProvided(), isVirtual(),
+          unless(anyOf(cxxConstructorDecl(), cxxDestructorDecl(),
+                       isCopyAssignmentOperator(), isMoveAssignmentOperator())))
+          .bind("virtual"),
+      this);
+  Finder->addMatcher(
+      cxxRecordDecl(has(accessSpecDecl(anyOf(isPrivate(), isProtected()))),
+                    isStruct())
+          .bind("access"),
+      this);
+  Finder->addMatcher(
+      cxxRecordDecl(hasDefinition(), isStruct(), hasBases()).bind("base"),
+      this);
+  Finder->addMatcher(
+      cxxRecordDecl(hasDefinition(), isStruct(), unless(isFinal()))
+          .bind("final"),
+      this);
+}
+
+void StructDefCheck::check(const MatchFinder::MatchResult &Result) {
+  auto Mgr = Result.SourceManager;
+
+  auto StructDef = Result.Nodes.getNodeAs<CXXMethodDecl>("special");
```

Comment:
This and the "virtual" matcher will match any method decl, even if they aren't defined in a struct. You could add a `hasParent(isStruct())` to these two matchers that checks that the recorddecl parent is a struct